File: src/test/java/com/lffblk/clone/model/AbstractObject.java
package com.lffblk.clone.model;

import java.util.Objects;
import java.util.Random;

abstract class AbstractObject {
    private AbstractObject yetAnotherItselfLink;
    private final int randomInt;
    private static int randomStaticField;

    AbstractObject() {
        this.yetAnotherItselfLink = this;
        Random random = new Random();
        this.randomInt = random.nextInt();
        randomStaticField = random.nextInt();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null) {
            return false;
        }
        if (getClass() != o.getClass()) {
            return false;
        }
        AbstractObject object = (AbstractObject) o;
        return this == yetAnotherItselfLink
            && object.randomInt == randomInt;
    }

    @Override
    public int hashCode() {
        return Objects.hash(randomInt, randomStaticField);
    }
}

File: src/test/java/com/lffblk/clone/model/AbstractObject.kt
package com.lffblk.clone.model

import java.util.Objects
import java.util.Random

abstract class AbstractObject {
    private val yetAnotherItselfLink: AbstractObject = this
    private val randomInt: Int

    init {
        val random = Random()
        this.randomInt = random.nextInt()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null) {
            return false
        }
        if (javaClass != other.javaClass) {
            return false
        }
        val `object` = other as AbstractObject?
        return this === yetAnotherItselfLink && `object`!!.randomInt == randomInt
    }

    override fun hashCode(): Int {
        return Objects.hash(randomInt)
    }
}
